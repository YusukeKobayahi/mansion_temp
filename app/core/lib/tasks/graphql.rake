# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
