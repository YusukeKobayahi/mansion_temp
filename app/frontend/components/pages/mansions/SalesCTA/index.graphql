fragment MsLibraryMansionsSalesCTA on Mansion {
  name
  street {
    id
  }
  blockNumber
}
