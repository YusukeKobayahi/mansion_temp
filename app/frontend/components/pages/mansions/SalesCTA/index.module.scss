@import "~/styles/variables";
@import "~/styles/utils";

.container {
  max-width: 750px;
  margin: -10px auto 0;
  color: $color-text-main;

  @include screenMd {
    width: 100%;
  }
}

.btn {
  display: block;
  margin: 0 auto;
  width: 500px;
  border: 2px solid #f15e3a;
  border-radius: 40px;
  font-size: 22px;
  color: #f15e3a;
  cursor: pointer;
  text-align: center;
  line-height: 56px;
  position: relative;
  font-weight: bold;
  background-color: #ffffff;

  @include screenMd {
    width: 100%;
    padding: 0;
    line-height: 43px;
    font-size: 12px;
    text-align: center;
  }

  &::after {
    content: "";
    position: absolute;
    top: 50%;
    right: 30px;
    width: 10px;
    height: 10px;
    border-top: 2px solid #f15e3a;
    border-right: 2px solid #f15e3a;
    transform: translateY(-50%) rotate(45deg);

    @include screenMd {
      right: 20px;
      width: 7px;
      height: 7px;
      border-top-width: 1px;
      border-right-width: 1px;
    }
  }
}
