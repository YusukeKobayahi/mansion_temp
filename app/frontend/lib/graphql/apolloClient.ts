// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
