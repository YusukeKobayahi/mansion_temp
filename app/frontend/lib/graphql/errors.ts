export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
