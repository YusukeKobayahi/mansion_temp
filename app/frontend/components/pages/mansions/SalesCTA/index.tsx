import { MsLibraryMansionsSalesCtaFragment } from "~/graphql/generated";
import { isSP } from "~/lib/utils";
import { event } from "~/lib/gtag";
import styles from "./index.module.scss";

type SalesCTAProps = {
  data: MsLibraryMansionsSalesCtaFragment;
};

const SalesCTA: React.FC<SalesCTAProps> = ({ data }: SalesCTAProps) => {
  const transitionToSalesForm = (): void => {
    event({
      action: "msl_cta_click",
      category: "msl_shosai",
      label: "msl_satei",
      value: 1,
    });
    const device = isSP() ? "sp" : "pc";

    location.href = `${
      process.env.NEXT_PUBLIC_SALES_FORM_URL
    }/lps/${device}/9_cl?${buildSalesFormParams()}`;
  };

  const buildSalesFormParams = (): string => {
    const utmParams = `utm_source=iel&utm_medium=notprovided&utm_campaign=mscta&utm_content=${location.href}`;
    const typeParams = `property-type-id=1&entry-type=mansion`;
    const streetId = data.street?.id || "";
    const blockNumber = data.blockNumber || "";
    const dataParams = `street-id=${streetId}&address=${blockNumber}&mansion-name=${data.name}`;

    return [utmParams, typeParams, dataParams].join("&");
  };

  return (
    <section className={styles.container}>
      <button className={styles.btn} onClick={transitionToSalesForm}>
        いますぐ無料査定を依頼する
      </button>
    </section>
  );
};

export default SalesCTA;
