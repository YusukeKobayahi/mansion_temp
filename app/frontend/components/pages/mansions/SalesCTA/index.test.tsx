import { render } from "@testing-library/react";
import mockData from "../fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import SalesCTA from ".";

describe("components/pages/mansions/SalesCTA", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(<SalesCTA data={mock.mansion} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
