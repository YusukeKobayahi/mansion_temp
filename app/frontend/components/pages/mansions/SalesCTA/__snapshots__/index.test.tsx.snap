// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/mansions/SalesCTA 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <button
      class="btn"
    >
      いますぐ無料査定を依頼する
    </button>
  </section>
</DocumentFragment>
`;
